water/lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -57,51 +61,70 @@ make1DTransferCommonAttrs(MemRefType memrefType, int64_t vectorizedDim,
 /// Materialize affine.apply for expressions inside a `map` with `symbols`.
 /// Each symbol is either a GPU id (thread/block) or a constant from `hyper`.
 static FailureOr<SmallVector<Value>>
-materializeAffine(Location loc, ArrayRef<wave::WaveSymbolAttr> symbols,
-                  AffineMap map, PatternRewriter &rewriter,
+materializeAffine(Location loc, ArrayRef<Attribute> symbols, AffineMap map,
+                  PatternRewriter &rewriter,
                   wave::WaveHyperparameterAttr hyper) {
   // NOTE: This helper assumes 0 dims in `map`. If you add dims, prepend
   // the dim operands before the symbol operands below.
   assert(map.getNumDims() == 0 && "expected 0 dims");
 
   auto threadId = [&](gpu::Dimension d) -> Value {
-    return rewriter.create<gpu::ThreadIdOp>(loc, rewriter.getIndexType(), d);
+    return gpu::ThreadIdOp::create(rewriter, loc, rewriter.getIndexType(), d);
   };
   auto blockId = [&](gpu::Dimension d) -> Value {
-    return rewriter.create<gpu::BlockIdOp>(loc, rewriter.getIndexType(), d);
+    return gpu::BlockIdOp::create(rewriter, loc, rewriter.getIndexType(), d);
   };
 
   SmallVector<Value> baseSymVals;
   baseSymVals.reserve(map.getNumSymbols());
-  int64_t numSym = map.getNumSymbols();
-  for (int64_t i = 0; i < numSym; ++i) {
-    StringRef name = symbols[i].getName();
-
-    Value v;
-    if (name == "_T0")
-      v = threadId(gpu::Dimension::x);
-    else if (name == "_T1")
-      v = threadId(gpu::Dimension::y);
-    else if (name == "_T2")
-      v = threadId(gpu::Dimension::z);
-    else if (name == "_WG0")
-      v = blockId(gpu::Dimension::x);
-    else if (name == "_WG1")
-      v = blockId(gpu::Dimension::y);
-    else if (name == "_WG2")
-      v = blockId(gpu::Dimension::z);
-    else if (llvm::is_contained({"_DD0", "_DD1", "_DD2"}, name))
-      return rewriter.notifyMatchFailure(
-          loc, "materialization of affine expressions containing device "
-               "dimension symbols is not implemented.");
-    else if (std::optional<int64_t> value = hyper.getSymbolValue(name)) {
-      v = rewriter.create<arith::ConstantIndexOp>(loc, *value);
-    } else {
-      LLVM_DEBUG(llvm::errs() << "symbol: " << name << "\n");
-      assert(false && "unknown symbol, should have been caught by verifiers");
+  for (Attribute attr : symbols) {
+    if (auto symbol = dyn_cast<wave::WaveSymbolAttr>(attr)) {
+      StringRef name = symbol.getName();
+      std::optional<int64_t> value = hyper.getSymbolValue(name);
+      if (!value) {
+        LLVM_DEBUG(llvm::errs() << "symbol: " << name << "\n");
```

Comment:
Nit: I'd rather use `#ifndef NDEBUG` around the entire if. `LLVM_DEBUG` is controlled by a different flag and is usually used in conjunction with `llvm::dbgs`.